### ISSUE TITLE
[Question] Parent Element in Layout upsets absolutely positioned children?

### ISSUE THREAD

**Issue by darth-cheney** [pharo-graphics/Bloc#16]
Hey guys,

I'm experiencing an issue and I don't know if it is an actual problem with the system or instead with my wielding of the tools.
  
Here's the rub: I started work on [trying to draw Copland/OS8 style windows using Bloc](https://github.com/darth-cheney/bloc-tests). I was able to come up with something that looks right, but uses crappy code and in which everything is absolutely positioned (no layouts).
  
If you pull down the latest from [the repo](https://github.com/darth-cheney/bloc-tests) and run:
```smalltalk
space := BlSpace new extent: 254@261; show.

win := CopWindowInner new.
space root addChild: win.
```
  
You should see something like:
![screen shot 2017-03-21 at 7 26 13 pm](https://cloud.githubusercontent.com/assets/1640299/24175386/49de9e26-0e6c-11e7-823f-81f131652301.png)
  
Which is getting pretty close already.
  
But now I've begun trying to actually get these titlebar elements into a true Bloc layout -- and here's where I'm running into some trouble.
  
`CopWindowButton` is a BlElement with several absolutely positioned children. These are very subtle (I copied them directly from an old OS8 screenshot) and precisely placed. They look great in the above example, but if I try to render a `CopWindowButton` instead in a parent who has a linear layout, the absolute positioned elements inside of the button are not updating themselves appropriately.
  
To make this clear, here are some examples:
  
### Absolute Positioning (works) ###
This example:
```smalltalk
space := BlSpace new extent:300@300; show.

button := CopWindowButton new.
button
	position: 2@2.

barColor := (Color r: 204 g: 204 b: 204 range: 255).

bar := BlElement new.
bar
	size: 100@18;
	background: barColor;
	position: 100@150.
bar addChild: button.
space root addChild: bar.
```
  
Gives you something like:
![screen shot 2017-03-21 at 7 33 00 pm](https://cloud.githubusercontent.com/assets/1640299/24175559/369ad86a-0e6d-11e7-9004-d0a854ceeb4b.png)
  
Which is clear and looks right. Zoom in on the image to see that the inner Blocs are  lined up correctly.
  
### Layout Positioning (not working right) ###
This example:
```smalltalk
space := BlSpace new extent:300@300; show.

button := CopWindowButton new.
button
	constraintsDo: [ :c | c linear vertical alignCenter ];
	margin: (BlInsets left: 2).

barColor := (Color r: 204 g: 204 b: 204 range: 255).

bar := BlElement new.
bar
	size: 100@18;
	position: 100@100;
	background: barColor;
	layout: (BlLinearLayout horizontal).
bar addChild: button.
space root addChild: bar.
```

Results in a hazier image:
![screen shot 2017-03-21 at 7 37 30 pm](https://cloud.githubusercontent.com/assets/1640299/24175675/dea0b624-0e6d-11e7-9069-9ae2e96433cc.png)
  
If you zoom in you'll see why: the inner Blocs in the button are not positioned absolutely within the `CopWindowButton`.
  
Here is a direct comparison, the first on top and the second on bottom, zoomed:
![comparison](https://cloud.githubusercontent.com/assets/1640299/24175768/71285ace-0e6e-11e7-9223-4a8212c137c5.png)

  
Any ideas?
  
Also, as an aside, is it OK for me to post my questions here on Github as I go, or is that annoying?

**Comment by syrel** [comment]
Hey Eric,

Yes, it is totally fine to ask here :) We learn from your experience and it will definitely help us make Bloc better!

The effect you see here is a result of antialias that happens when elements have float positions. Bloc is a vector graphics framework and therefore allows elements to have non-integer coordinates. More precisely, Bloc does not operate in terms if pixels instead all numbers you specify are measured in density independent pixels (dip or just dp), so that 1dp = 1px on PC screens with 96 dots per inch. On mac retina displays 1dp = 2px. On Android 1dp = 1px on 160dpi screens.

Below are sketches of your two cases:

In the first example you rely on absolute positioning, where element coordinates are set by a user. In your case it is (2dp@2dp). Note that since bar has height 18dp and button 13dp we get free space below the button of height 3dp.

<img width="766" alt="screen shot 2017-03-22 at 09 35 23" src="https://cloud.githubusercontent.com/assets/5487042/24188783/f8909386-0ee2-11e7-9af5-80eb865d0b9d.png">

In the second example you let linear layout to put your button exactly in the middle on the bar and have left margin equal to 2dp. In this case layout computes overall free space which is 18dp - 13dp = 5dp and then divides it by 2 to get top offset: 5dp / 2 = 2.5dp. Then during rendering process on non-retina screen (Pharo assumes screen dpi is 96) underlying engine has to render a button at coordinate (2dp@2.5dp) * 1px/1dp => (2px@2.5px) that results in blurry picture. Note, that on retina screen resulting position in device space coordinates would be (2dp@2.5dp) * 2px / 1dp => (4px@5px) which would give us an amazing quality.

<img width="766" alt="screen shot 2017-03-22 at 09 35 32" src="https://cloud.githubusercontent.com/assets/5487042/24188784/f9dad24c-0ee2-11e7-9c81-f4b13880659f.png">

This example reveals one of the disadvantages of the *vector* graphical framework.

**Comment by darth-cheney** [comment]
Excellent explanation @syrel , thanks. I think the immediate solution is either to  keep buttons like this absolutely positioned (they don't really need to move around since they'll always be in the same spot) or to extend the height of the bar by 1 so that the spacing is even above and below.
  
This brings up another issue I was having: if the button is in a linear layout, `position:` has no effect. There for I instead tried to set `margin: (BlInsets top: 2 left: 2)1, but this only seems to set the margin on the left and not on the top. Am I also using this incorrectly?
  
I also have a couple of related questions:
1.  Is there a way to bind a child bloc to a specific size dimension (w or h) of its parent bloc without using a layout?
2.  In a linear layout is there currently a way to have some elements "flex" to fill in unused horizontal space while others have an absolutely specified width? The examples in my case are the "slats" that display in the titlebar, the lines adjacent to buttons. Buttons do not need to flex (blue) but the slats (red) need to stretch to fill space unused by either the buttons or the title (green):
![topbarannotated](https://cloud.githubusercontent.com/assets/1640299/24216922/6202975c-0f14-11e7-9c14-731aede0a980.png)
3. What exactly is the difference between `extent:` and `size:`?

Thanks again

**Comment by syrel** [comment]
Yes, `position:` has no effect if custom layout is applied. That is the intent of a layout to position elements :) Btw, `BlFrameLayout` is designed specifically for absolute positioning + it additionally supports alignment strategies (left, top, center etc).

> margin: (BlInsets top: 2 right: 0 bottom: 0 left: 2)

Could you post a reproducible example that shows that it does not work? Because linear layout has quite good test coverage. Maybe that it is a rare case that was not covered...

> Is there a way to bind a child bloc to a specific size dimension (w or h) of its parent bloc without using a layout?

Yes, it is possible :) Here is an example of a blue ellipse matching the size of a red rectangular parent:

```
BlElement new
   size: 300@200;
   background: (Color red alpha: 0.2);
   addChild: (BlElement new
      constraintsDo: [ :c |
         c horizontal matchParent.
         c vertical matchParent ];
      geometry: BlEllipse new;
      background: (Color blue alpha: 0.4))
```

![match-parent](https://cloud.githubusercontent.com/assets/5487042/24220604/2677da94-0f4b-11e7-923c-06933cb210c6.png)

The idea is to tell a child that it should `matchParent` horizontally or vertically.

**Note:** Default layout of the element does not support padding and margin. For absolute positioning, margin, padding, stretching I recommend to use `BlFrameLayout`. It was designed specifically for that case.

> In a linear layout is there currently a way to have some elements "flex" to fill in unused horizontal space while others have an absolutely specified width? The examples in my case are the "slats" that display in the titlebar, the lines adjacent to buttons. Buttons do not need to flex (blue) but the slats (red) need to stretch to fill space unused by either the buttons or the title (green):

I prepared an example that uses `BlGridLayout` in order to have equal spacing between all elements and parent (`cellSpacing:`). The same can be archived with the help of `BlLinearLayout` with the only difference that margin/padding should be set manually for every child element.

Copy link, paste it in Spotter (shift+enter) in Pharo and press enter, will open Playground with code snippet:
[http://ws.stfx.eu/KEFZ437G1DOG](http://ws.stfx.eu/KEFZ437G1DOG)
![bloc-title-example](https://cloud.githubusercontent.com/assets/5487042/24220767/eb453272-0f4b-11e7-91e6-f2406aa6e658.png)

> What exactly is the difference between extent: and size:?

`extent:` changes extent property directly and intended for internal use only. Extent property is only valid until next layout pass (next frame) => temporary.
`size:` changes layout constraints and requests a layout update on next frame which will result in `extent:`. Layout constraints should be used by users to **declare** UI layout => permanent:

**Note:** All layout constraints changes are accumulated during a frame and applied only on the next frame. It allows us to compute layout only once per frame and not after every change.

**Comment by darth-cheney** [comment]
Awesome, thanks @syrel. I'm going to play with this some more later this evening.
  
In the meantime, here's the example of the top margin in linear layout not working as (I assume) it should:
```smalltalk
space := BlSpace new extent: 300@300; show.

barColor := (Color r: 204 g: 204 b: 204 range: 255).

bar := BlElement new.
bar
	size: 100@18;
	layout: (BlLinearLayout horizontal);
	background: barColor.
bar position: 100@100.

button := CopWindowButton new.
button
	margin: (BlInsets top: 2 right: 0 bottom: 0 left: 2).
bar addChild: button.

space root addChild: bar.
```
  
Which produces:
![screen shot 2017-03-22 at 5 19 43 pm](https://cloud.githubusercontent.com/assets/1640299/24221205/c56fab36-0f23-11e7-8c2d-d0b340f7f87b.png)

**Comment by syrel** [comment]
Perfect! It indeed does not work :)
Thanks for this bug, I will take a look

**Comment by darth-cheney** [comment]
Another somewhat related question.
Though I can set the width (horizontal) or height (vertical) of an element to match its parent, is there some way to set it to a computation of the same value in the parent? For example, instead of matching a parent horizontal 100%, can I match it 80%?
  
The reason I ask is because of the "slats" object. Slats look like this in the source images:
![coplandslatscloseup](https://cloud.githubusercontent.com/assets/1640299/24265177/035b84fa-0fd9-11e7-89fd-5502a7d354c1.png)
  
You can see that it's made up of light and dark bars alternating vertically, where the light bars end 1 or 1 before the right and the dark bars start 1 or 2 from the left. 
  
I've tried the below code in order to achieve the same effect, but it seems to just create the bars full width:
```smalltalk
initialize
	super initialize.
	self layout: (BlLinearLayout vertical).
	self constraintsDo: [ :c |
		c vertical matchParent ].
	self drawSlats.
	^ self.
```
```smalltalk
drawSlats
	"Draws alternating light and dark slats,
	 starting with light, with alternating
	 margins on either side."

	| darkColor lightColor |
	darkColor := (Color gray lighter lighter).
	lightColor := Color white darker.
	
	6 timesRepeat: [ 
		self addChild: (BlElement new 
			margin: (BlInsets right: 2);
			constraintsDo: [ :c |
				c vertical exact: 1.
				c horizontal matchParent ];
			background: lightColor).
		self addChild: (BlElement new
			margin: (BlInsets left: 2);
			constraintsDo: [ :c | 
				c vertical exact: 1.
				c horizontal matchParent ];
			background: darkColor).
	].
```
Is this related to the top margin bug or am I doing something wrong/not possible?

**Comment by syrel** [comment]
> Another somewhat related question.

Questions are always welcomed :)

> For example, instead of matching a parent horizontal 100%, can I match it 80%?

Yes, that is possible! Linear layout supports it by using `weight` parameter.

By default, all children matching parent has `weight: 1.0` that is why if added to a parent with horizontal layout each takes 50% of parent's size.
![bloc-layout-weight 001](https://cloud.githubusercontent.com/assets/5487042/24289176/6a6932da-1080-11e7-8f8f-11dabae917a4.png)

Weight can be customised by sending `weight:` message with a new value as a parameter to linear layout constraints:

```smalltalk
child constraintsDo: [ :c | c linear weight: 0.5 ]
```

This allows us to archive a proportional layout:
![bloc-layout-weight 002](https://cloud.githubusercontent.com/assets/5487042/24289321/0a208670-1081-11e7-9173-af2ba3cb3d95.png)

However, by default linear layout has undefined weight sum, which is computed on fly, that prevents a single child from being proportionally layered out. For that we have to set it manually:

![bloc-layout-weight 003](https://cloud.githubusercontent.com/assets/5487042/24289391/65e84754-1081-11e7-921c-4d93ab2f72c1.png)

Hope this helps :) Have fun!

**Comment by dionisiydk** [comment]
That is interesting approach :)

2017-10-18 12:22 GMT+02:00 Aliaksei Syrel <notifications@github.com>:
